File: Assets/Bundles/Server/Resources/TS/Services/Global/Generator/GeneratorService.lua
-- Compiled with unity-ts v2.1.0-75
local TS = require("Shared/include/RuntimeLib")
local _flamework_core = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init")
local Reflect = _flamework_core.Reflect
local Flamework = _flamework_core.Flamework
local Service = require("Shared/rbxts_include/node_modules/@easy-games/flamework-core/out/init").Service
local ObjectUtil = require("Shared/rbxts_include/node_modules/@easy-games/unity-object-utils/init")
local ServerSignals = require("Server/TS/ServerSignals").ServerSignals
local CharacterEntity = require("Shared/TS/Entity/Character/CharacterEntity").CharacterEntity
local ItemStack = require("Shared/TS/Inventory/ItemStack").ItemStack
local Network = require("Shared/TS/Network").Network
local Task = require("Shared/TS/Util/Task").Task
local TimeUtil = require("Shared/TS/Util/TimeUtil").TimeUtil
local SetInterval = require("Shared/TS/Util/Timer").SetInterval
-- * Default generator stack size. We _always_ start at 0.
local DEFAULT_GENERATOR_STACK_SIZE = 0
-- * Snapshot send delay after user connects.
local SNAPSHOT_SEND_DELAY = 1
-- * Generator item spawn offset. Items spawn _above_ generators and fall to the ground.
local GENERATOR_ITEM_SPAWN_OFFSET = Vector3.new(0, 2.8, 0)
local GeneratorService
do
	GeneratorService = setmetatable({}, {
		__tostring = function()
			return "GeneratorService"
		end,
	})
	GeneratorService.__index = GeneratorService
	function GeneratorService.new(...)
		local self = setmetatable({}, GeneratorService)
		return self:constructor(...) or self
	end
	function GeneratorService:constructor()
		self.generatorIdCounter = 0
		self.generatorMap = {}
		self.stackMap = {}
	end
	function GeneratorService:OnStart()
		-- Split resources
		ServerSignals.EntityPickupItem:Connect(function(event)
			local groundObjectAttributes = event.groundItemGO:GetComponent("EasyAttributes")
			local generatorId = groundObjectAttributes:GetString("generatorId")
			if not (generatorId ~= "" and generatorId) then
				return nil
			end
			local genState = self:GetGeneratorById(generatorId)
			if not genState then
				return nil
			end
			local pickupPlayer = event.entity.player
			local _condition = pickupPlayer
			if _condition then
				local _result = genState
				if _result ~= nil then
					_result = _result.split
				end
				_condition = _result
			end
			if _condition then
				local splitTeam = pickupPlayer:GetTeam()
				if not splitTeam then
					return nil
				end
				local splitRange = genState.split.range
				local _exp = splitTeam:GetPlayers()
				local _arg0 = function(player)
					local playerEntity = (Flamework.resolveDependency("Bundles/Server/Services/Global/Entity/EntityService@EntityService")):GetEntityByClientId(player.clientId)
					if not playerEntity then
						return nil
					end
					if not (TS.instanceof(playerEntity, CharacterEntity)) then
						return nil
					end
					local _position = playerEntity.gameObject.transform.position
					local _pos = genState.dto.pos
					local distanceFromGen = (_position - _pos).magnitude
					if player ~= pickupPlayer and distanceFromGen <= splitRange then
						local inv = playerEntity:GetInventory()
						inv:AddItem(ItemStack.new(event.itemStack:GetItemType(), event.itemStack:GetAmount()))
					end
				end
				for _v in _exp do
					_arg0(_v, _v, _exp)
				end
			end
			-- Generator cleanup.
			genState.stackSize = 0
			self.stackMap[generatorId] = nil
		end)
		-- Handle late joiners.
		ServerSignals.PlayerJoin:connect(function(event)
			Task:Delay(SNAPSHOT_SEND_DELAY, function()
				local _fn = Network.ServerToClient.GeneratorSnapshot.Server
				local _exp = event.player.clientId
				local _exp_1 = self:GetAllGenerators()
				local _arg0 = function(state)
					return state.dto
				end
				-- ▼ ReadonlyArray.map ▼
				local _newValue = table.create(#_exp_1)
				for _k, _v in _exp_1 do
					_newValue[_k] = _arg0(_v, _k - 1, _exp_1)
				end
				-- ▲ ReadonlyArray.map ▲
				_fn:FireClient(_exp, _newValue)
			end)
		end)
	end
	function GeneratorService:CreateGenerator(pos, config)
		local generatorId = self:GenerateGeneratorId()
		local state = {
			stackLimit = config.stackLimit,
			stackSize = 0,
			originalSpawnRate = config.spawnRate,
			dto = {
				pos = pos,
				id = generatorId,
				item = config.item,
				spawnRate = config.spawnRate,
				nextSpawnTime = TimeUtil:GetServerTime() + config.spawnRate,
				label = config.label,
			},
		}
		state.ticker = self:TickGenerator(state)
		-- Store generator in map, notify client of generator creation.
		local _generatorMap = self.generatorMap
		local _id = state.dto.id
		_generatorMap[_id] = state
		Network.ServerToClient.GeneratorCreated.Server:FireAllClients(state.dto)
		-- Return id.
		return generatorId
	end
	function GeneratorService:GetGeneratorById(generatorId)
		local _generatorMap = self.generatorMap
		local _generatorId = generatorId
		return _generatorMap[_generatorId]
	end
	function GeneratorService:GenerateGeneratorId()
		local generatorId = "generator_" .. tostring(self.generatorIdCounter)
		self.generatorIdCounter += 1
		return generatorId
	end
	function GeneratorService:TickGenerator(generatorState)
		return SetInterval(generatorState.dto.spawnRate, function()
			-- Always update next spawn time.
			generatorState.dto.nextSpawnTime = TimeUtil:GetServerTime() + generatorState.dto.spawnRate
			-- Only increase stack size if generator has _not_ reached capacity.
			if generatorState.stackSize < generatorState.stackLimit then
				generatorState.stackSize += 1
				local _stackMap = self.stackMap
				local _id = generatorState.dto.id
				local existingStack = _stackMap[_id]
				if existingStack then
					existingStack:SetAmount(generatorState.stackSize)
				else
					local newGeneratorStack = ItemStack.new(generatorState.dto.item, 1)
					local _stackMap_1 = self.stackMap
					local _id_1 = generatorState.dto.id
					_stackMap_1[_id_1] = newGeneratorStack;
					(Flamework.resolveDependency("Bundles/Server/Services/Global/GroundItem/GroundItemService@GroundItemService")):SpawnGroundItem(newGeneratorStack, generatorState.dto.pos + GENERATOR_ITEM_SPAWN_OFFSET, nil, generatorState.dto.id)
				end
			end
		end)
	end
	function GeneratorService:UpdateGeneratorSpawnRateById(generatorId, newSpawnRate)
		local _generatorMap = self.generatorMap
		local _generatorId = generatorId
		local state = _generatorMap[_generatorId]
		if not state then
			return nil
		end
		-- Stop ticker and reconstruct dto with updated state.
		local _result = state.ticker
		if _result ~= nil then
			_result()
		end
		state.dto.spawnRate = newSpawnRate
		state.dto.nextSpawnTime = TimeUtil:GetServerTime() + newSpawnRate
		state.ticker = self:TickGenerator(state)
		-- Inform clients of _all_ server-sided generator spawn rate changes.
		Network.ServerToClient.GeneratorSpawnRateChanged.Server:FireAllClients(state.dto.id, newSpawnRate)
	end
	function GeneratorService:GetAllGenerators()
		return ObjectUtil.values(self.generatorMap)
	end
end
-- (Flamework) GeneratorService metadata
Reflect.defineMetadata(GeneratorService, "identifier", "Bundles/Server/Services/Global/Generator/GeneratorService@GeneratorService")
Reflect.defineMetadata(GeneratorService, "flamework:implements", { "$:flamework@OnStart" })
Reflect.decorate(GeneratorService, "$:flamework@Service", Service, { {} })
return {
	GeneratorService = GeneratorService,
}
-- ----------------------------------
-- ----------------------------------

File: Assets/Typescript~/src/Server/Services/Global/Generator/GeneratorService.ts
import { Dependency, OnStart, Service } from "@easy-games/flamework-core";
import ObjectUtil from "@easy-games/unity-object-utils";
import { ServerSignals } from "Server/ServerSignals";
import { CharacterEntity } from "Shared/Entity/Character/CharacterEntity";
import { GeneratorCreationConfig } from "Shared/Generator/GeneratorMeta";
import { ItemStack } from "Shared/Inventory/ItemStack";
import { Network } from "Shared/Network";
import { Task } from "Shared/Util/Task";
import { TimeUtil } from "Shared/Util/TimeUtil";
import { SetInterval } from "Shared/Util/Timer";
import { EntityService } from "../Entity/EntityService";
import { GroundItemService } from "../GroundItem/GroundItemService";
import { GeneratorState } from "./GeneratorState";
/** Default generator stack size. We _always_ start at 0.*/
const DEFAULT_GENERATOR_STACK_SIZE = 0;
/** Snapshot send delay after user connects. */
const SNAPSHOT_SEND_DELAY = 1;
/** Generator item spawn offset. Items spawn _above_ generators and fall to the ground. */
const GENERATOR_ITEM_SPAWN_OFFSET = new Vector3(0, 2.8, 0);

@Service({})
export class GeneratorService implements OnStart {
	/** Generator id counter. */
	private generatorIdCounter = 0;
	/** Mapping of generator id to `GeneratorStateDto`. */
	private generatorMap = new Map<string, GeneratorState>();
	/** Mapping of generator id to `ItemStack`. */
	private stackMap = new Map<string, ItemStack>();

	constructor() {}

	OnStart(): void {
		// Split resources
		ServerSignals.EntityPickupItem.Connect((event) => {
			const groundObjectAttributes = event.groundItemGO.GetComponent<EasyAttributes>();
			const generatorId = groundObjectAttributes.GetString("generatorId");
			if (!generatorId) return;

			const genState = this.GetGeneratorById(generatorId);
			if (!genState) return;
			const pickupPlayer = event.entity.player;
			if (pickupPlayer && genState?.split) {
				const splitTeam = pickupPlayer.GetTeam();
				if (!splitTeam) return;
				const splitRange = genState.split.range;
				splitTeam.GetPlayers().forEach((player) => {
					const playerEntity = Dependency<EntityService>().GetEntityByClientId(player.clientId);
					if (!playerEntity) return;
					if (!(playerEntity instanceof CharacterEntity)) return;
					const distanceFromGen = playerEntity.gameObject.transform.position.sub(genState.dto.pos).magnitude;
					if (player !== pickupPlayer && distanceFromGen <= splitRange) {
						const inv = playerEntity.GetInventory();
						inv.AddItem(new ItemStack(event.itemStack.GetItemType(), event.itemStack.GetAmount()));
					}
				});
			}
			/* Generator cleanup. */
			genState.stackSize = 0;
			this.stackMap.delete(generatorId);
		});

		/* Handle late joiners. */
		ServerSignals.PlayerJoin.Connect((event) => {
			Task.Delay(SNAPSHOT_SEND_DELAY, () => {
				Network.ServerToClient.GeneratorSnapshot.Server.FireClient(
					event.player.clientId,
					this.GetAllGenerators().map((state) => state.dto),
				);
			});
		});
	}

	/**
	 * Creates a generator at `generatorPosition` based on `generatorConfig`.
	 * @param pos The position to create generator at.
	 * @param config Generator config.
	 * @returns Generator id.
	 */
	public CreateGenerator(pos: Vector3, config: GeneratorCreationConfig): string {
		const generatorId = this.GenerateGeneratorId();

		const state: GeneratorState = {
			stackLimit: config.stackLimit,
			stackSize: 0,
			originalSpawnRate: config.spawnRate,
			dto: {
				pos: pos,
				id: generatorId,
				item: config.item,
				spawnRate: config.spawnRate,
				nextSpawnTime: TimeUtil.GetServerTime() + config.spawnRate,
				label: config.label,
			},
		};
		state.ticker = this.TickGenerator(state);

		/* Store generator in map, notify client of generator creation. */
		this.generatorMap.set(state.dto.id, state);
		Network.ServerToClient.GeneratorCreated.Server.FireAllClients(state.dto);
		/* Return id. */
		return generatorId;
	}

	/**
	 * Fetch a generator by generator id.
	 * @param generatorId A generator id.
	 * @returns Generator state dto.
	 */
	public GetGeneratorById(generatorId: string): GeneratorState | undefined {
		return this.generatorMap.get(generatorId);
	}

	/** Generates and returns a unique generator id. */
	private GenerateGeneratorId(): string {
		const generatorId = `generator_${this.generatorIdCounter}`;
		this.generatorIdCounter++;
		return generatorId;
	}

	/** Ticks a generator on server. */
	private TickGenerator(generatorState: GeneratorState): () => void {
		return SetInterval(generatorState.dto.spawnRate, () => {
			/* Always update next spawn time. */
			generatorState.dto.nextSpawnTime = TimeUtil.GetServerTime() + generatorState.dto.spawnRate;
			/* Only increase stack size if generator has _not_ reached capacity. */
			if (generatorState.stackSize < generatorState.stackLimit) {
				generatorState.stackSize++;

				const existingStack = this.stackMap.get(generatorState.dto.id);
				if (existingStack) {
					existingStack.SetAmount(generatorState.stackSize);
				} else {
					const newGeneratorStack = new ItemStack(generatorState.dto.item, 1);
					this.stackMap.set(generatorState.dto.id, newGeneratorStack);
					Dependency<GroundItemService>().SpawnGroundItem(
						newGeneratorStack,
						generatorState.dto.pos.add(GENERATOR_ITEM_SPAWN_OFFSET),
						undefined,
						generatorState.dto.id,
					);
				}
			}
		});
	}

	/**
	 * Set generator spawn rate based on generator id.
	 * @param generatorId A generator id.
	 * @param newSpawnRate New spawn rate in seconds.
	 */
	public UpdateGeneratorSpawnRateById(generatorId: string, newSpawnRate: number): void {
		const state = this.generatorMap.get(generatorId);
		if (!state) return;

		// Stop ticker and reconstruct dto with updated state.
		state.ticker?.();

		state.dto.spawnRate = newSpawnRate;
		state.dto.nextSpawnTime = TimeUtil.GetServerTime() + newSpawnRate;
		state.ticker = this.TickGenerator(state);

		/* Inform clients of _all_ server-sided generator spawn rate changes. */
		Network.ServerToClient.GeneratorSpawnRateChanged.Server.FireAllClients(state.dto.id, newSpawnRate);
	}

	/**
	 * @returns All `GeneratorStateDto`s on server.
	 */
	public GetAllGenerators(): GeneratorState[] {
		return ObjectUtil.values(this.generatorMap);
	}
}
